### PR TITLE
Concave hull

### PR DESCRIPTION
**Pull request by FObermaier**
I hijacked @dr-jts `ConcaveHull` _stub_ in jts-lab and replaced it with an implementation inspired by
@mourner 's [concaveman](https://github.com/mapbox/concaveman) algorithm.

It is based on ideas from the paper _[A New Concave Hull Algorithm and Concaveness Measure for n-dimensional Datasets, 2012](http://www.iis.sinica.edu.tw/page/jise/2012/201205_10.pdf)_ by Jin-Seo Park and Se-Jong Oh.

### CODE REVIEW

**Comment by dr-jts** [comment]
Can this algorithm produce invalid topology for extreme tolerance values?

This is possibly one advantage of the triangulation-based approaches, which can be easily limited to ensure that only valid topology is produced.  And it's also easy to enforce fully connected output.

File: modules/core/src/main/java/org/locationtech/jts/algorithm/Distance.java [hunk 1]
```diff
@@ -219,4 +219,25 @@ public static double pointToLinePerpendicular(Coordinate p,
...
 
+  /**
+   * Computes the distance between the segment {@code A}{@code B} and the
+   * { @link Envelope } {@code bounds}
+   * @param A the starting point of the segment
+   * @param B the end point of the segment
+   * @param bounds the bounds
+   * @return the distance between AB and the envelope.
+   */
+  public static double segmentToEnvelope(Coordinate A, Coordinate B, Envelope bounds) {
```
[the remark this answers is not in the document]
Comment (FObermaier) [reply modules/core/src/main/java/org/locationtech/jts/algorithm/Distance.java]: Actually this seems to be a _left-over_ from the first attempts using real distances.
But yes, I can move it to `Envelope` class

File: modules/lab/src/main/java/org/locationtech/jts/hull/ConcaveHull.java [hunk 2]
```diff
@@ -11,43 +11,462 @@
...
+ * This implementation dramatically improves performance over the one stated in the paper (O(rn), where r is a
+ * number of output points, to O(n log n)) by introducing a fast k nearest points to a segment algorithm, a
+ * modification of a depth-first kNN R-tree search using a priority queue.
+ * </p>
+ *
+ * @since 1.17
+ * @author Felix Obermaier
+ */
+@SuppressWarnings({"ForLoopReplaceableByForEach", "ManualArrayToCollectionCopy"})
 public class ConcaveHull {
```
[the remark this answers is not in the document]
Comment (FObermaier) [reply modules/lab/src/main/java/org/locationtech/jts/hull/ConcaveHull.java]: Alternatives:
* `ConcaveHullDigger` (sounds like an arcade game)
* `ParkOhConcaveHull` (giving credit to the inventors)

Anyone?

**Comment by FObermaier** [comment]
> In most other algorithms I've seen (such as Duckhams - implementation here) larger tolerance produce a more concave hull.

I have no idea how to change that for the `lengthThreshold` parameter. It prevents looking for further candidates when a segment's length goes below that value. The `concavity` value could be handled as a reciprocal value thus having greater concaveness for larger values.

> Can this algorithm produce invalid topology for extreme tolerance values?

According to the short paper: 
>  Our proposed algorithm satisfies Galton’s evaluation criteria, with the exception of
question (5). The question has __no relationship to the robustness__ of the proposed algorithm.

Question 5 asks if the concave hull should be polygonal, or can its boundary be curved?

As the starting solution is a convex hull, I see no way of producing a not fully connected result, regardless of the input parameters.

**Comment by FObermaier** [comment]
Another idea to improve the _exprected_ output is to mark segments that are already part of the input geometry as immutable. The other concave hull implementation you mentioned proposed to `densify`  the input geometry.

File: modules/lab/src/main/java/org/locationtech/jts/hull/ConcaveHull.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (dr-jts) [reply modules/lab/src/main/java/org/locationtech/jts/hull/ConcaveHull.java]: I'd prefer a name based on the algorithm concept (unless this becomes known by the names of the inventors).  I agree Digger is an awkward term, but that's the one they use. Perhaps `DiggingConcaveHull` ?

File: modules/core/src/main/java/org/locationtech/jts/algorithm/Distance.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (dr-jts) [reply modules/core/src/main/java/org/locationtech/jts/algorithm/Distance.java]: If not used then just delete it.

**Comment by dr-jts** [comment]
> > In most other algorithms I've seen (such as Duckhams - implementation here) larger tolerance produce a more concave hull.
> 
> I have no idea how to change that for the `lengthThreshold` parameter. It prevents looking for further candidates when a segment's length goes below that value. 

Yes, this seems more like a quality control rather than a core parameter to the algorithm.  Was this in the original paper?

> The `concavity` value could be handled as a reciprocal value thus having greater concaveness for larger values.

Yes, I thought of that too.  Best to leave it as per original paper, though.

**Comment by dr-jts** [comment]
> As the starting solution is a convex hull, I see no way of producing a not fully connected result, regardless of the input parameters.

I guess the real issue is whether the algorithm can produce edges which cross one another.  My evaluation is that this is not possible, due to the closest edge-point pair always being chosen for processing.  Would have been nice if they'd proved this (it's not part of the Galton criteria, but probably should be).

File: modules/lab/src/main/java/org/locationtech/jts/hull/ConcaveHull.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (dr-jts) [reply modules/lab/src/main/java/org/locationtech/jts/hull/ConcaveHull.java]: Perhaps `NearestPointConcaveHull` ?

**Comment by dr-jts** [comment]
> This is possibly one advantage of the triangulation-based approaches, which can be easily limited to ensure that only valid topology is produced. And it's also easy to enforce fully connected output.

Triangulation-based approaches can also produce holes - which may or may not be desirable (but obviously can be removed if not).

**Comment by dr-jts** [comment]
> This relates to something I find problematic about current concave hull algorithms, which is that it's hard to predict a priori what a "good" tolerance distance is. Actually this algorithm might work better in this respect, since the tolerance is a ratio rather than an absolute distance.

Another comment on the intuitiveness of the concavity parameter is that (if I understand it) it's based on purely local information (a ratio of the edges of the triangle being considered for removal).  

It's interesting that they propose a **Concaveness Measure** equation which is based on global values (perimeter length). But there doesn't seem to be any way to decide _a priori_ on a concavity value that will produce a target concaveness measure value.

**Comment by FObermaier** [comment]
> Another possible issue is whether the algorithm can produce edges which cross one another.

That is taken care of:
https://github.com/FObermaier/jts/blob/3a8e9f4459c2deb76b3db6f32f98a691e3dd6b7b/modules/lab/src/main/java/org/locationtech/jts/hull/ConcaveHull.java#L341-L344

**Comment by FObermaier** [comment]
> Triangulation-based approaches can also produce holes - which may or may not be desirable (but obviously can be removed if not).

In my understanding hulls don't have holes.

File: modules/lab/src/main/java/org/locationtech/jts/hull/ConcaveHull.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (FObermaier) [reply modules/lab/src/main/java/org/locationtech/jts/hull/ConcaveHull.java]: Using _to excavate_ as a synonyme of _to dig_ gives a less awkward so how about `ConcaveHullExcavator`?

**Comment by FObermaier** [comment]
I'm wondering if there is an interest in this convex hull implemenation?

**Comment by FObermaier** [comment]
The code can be found here: https://github.com/FObermaier/jts-ex

**Comment by dr-jts** [comment]
I haven't forgotten about this, but need to do #823 as a drop-in replacement for the PostGIS one.

It would be interesting to compare the results and performance of the two algorithms.

**Pull request closed without merging**